lib/NMISNG/Auth.pm

```diff
@@ -2141,4 +2199,383 @@ sub update_last_login
+	if ( exists $UT->{$user}{privilege} and $UT->{$user}{privilege} ne ""  ) {
+		NMISNG::Util::logAuth("DEBUG Auth::_GetPrivs, User '$user' using local for privs") if $self->{debug};
+		$self->{priv} = $UT->{$user}{privilege};
+		$user_local = 1;
+		NMISNG::Util::logAuth("DEBUG Auth::_GetPrivs, User '$user' use priv: $self->{priv}") if $self->{debug};
+    } elsif ( $auth_ldap_privs && $auth_ldap_acc ) { # If 'auth_ldap_acc' is unset, then we are authenticating without authorization, so we have to assign default priveleges unless we are defined locally.
+		NMISNG::Util::logAuth("DEBUG Auth::_GetPrivs, User '$user' using _get_ldap_privs for privs") if $self->{debug};
+		($self->{priv}, $groupList) = $self->_get_ldap_privs(user => $user);
+		NMISNG::Util::logAuth("DEBUG Auth::_GetPrivs, User '$user' use priv: $self->{priv}") if $self->{debug};
+		NMISNG::Util::logAuth("DEBUG Auth::_GetPrivs, User '$user' use mapping for groups: $groupList") if $self->{debug};
+	}
+	else {
+		if ( $auth_default_privilege ne "" and !NMISNG::Util::getbool($auth_default_privilege,"invert") ) {
+			$self->{priv} = $auth_default_privilege;
+			$self->{privlevel} = 5;
+			NMISNG::Util::logAuth("WARN Auth::_GetPrivs, User '$user' not found in Users table, assigned default privilege '$auth_default_privilege.'");
+			$user_local = 1;
+		} else {
+			$self->{priv} = "";
+			$self->{privlevel} = 5;
+			NMISNG::Util::logAuth("ERROR Auth::_GetPrivs, User '$user' not found in Users table, no default privilege configured");
+			return 0;
+		}
+		# We dont have a user in NMIS but we want the user to be able to login and take default groups
+		if ( $auth_default_groups and $auth_default_groups ne "" ) {
+			$UT->{$user}{groups} = $auth_default_groups;
+		}
+	}
+
+	if ( ! exists $PMT->{$self->{priv}} and $PMT->{$self->{priv}}{level} eq "" ) {
+		NMISNG::Util::logAuth("WARN Auth::_GetPrivs, Privilege '$self->{priv}' not found for user '$user'.");
+		$self->{priv} = "";
+		$self->{privlevel} = 5;
+		return 0;
+	}
+
+	$self->{privlevel} = 5;
+	if ( $PMT->{$self->{priv}}{level} ne "" ) {
+		$self->{privlevel} = $PMT->{$self->{priv}}{level};
+	}
+	NMISNG::Util::logAuth("DEBUG Auth::_GetPrivs, User '$user' has priv='$self->{priv}' and privlevel='$self->{privlevel}'.") if $self->{debug};
+
+	if ($auth_ldap_privs and $user_local == 0) {
+		$self->SetGroups( rawgroups => $groupList );
+	} else {
+		$self->SetGroups( rawgroups => $UT->{$user}{groups} );
+	}
+   	NMISNG::Util::logAuth("DEBUG Auth::_GetPrivs, User '$user' has groups '" . join(", ", @{$self->{groups}}) . "'.") if ($self->{debug});
+	
+	return 1;
+}
+
+# Mapping with ldap
+# Get groups for the logged in user from ldap
+# And map the groups with a local privilege from nmis
+# Returns privilege, and group
+sub _get_ldap_privs
+{
+	my ($self, %args) = @_;
+	my $user = $args{"user"};
+	my $ldap;
+    my $sec = 0;
+	my $auth_ldap_base       = $self->{config}->{'auth_ldap_base'} // $self->{config}->{'auth_ldap_context'}; # retured value
```

Comment:
if this is the same as the code above please move them into a common function

Reply:
Resolved